lib/src/merged_tree.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::max;
+
+use crate::backend::TreeValue;
+use crate::conflicts::Conflict;
+use crate::repo_path::{RepoPath, RepoPathComponent};
+use crate::tree::Tree;
+use crate::tree_builder::TreeBuilder;
+
+/// Presents a view of a merged set of trees.
+pub enum MergedTree {
+    /// A single tree, possibly with path-level conflicts.
+    Legacy(Tree),
+    /// A merge of multiple trees, or just a single tree. The individual trees
+    /// have no path-level conflicts.
+    Merge(Conflict<Tree>),
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub enum MergedTreeValue<'a> {
+    Resolved(Option<&'a TreeValue>),
+    /// TODO: Make this a `Conflict<Option<&'a TreeValue>>` (reference to the
+    /// value) once we have removed the `MergedTree::Legacy` variant.
```

Comment:
I understood the issue here: it's whether any `TreeValue::Tree` values can have conflicts or not.

I think you should mention this in a comment.

More optionally (haven't quite thought this through), perhaps we could rename `Resolved` and `Conflict` to `Legacy` (or `LegacyTree`) and `Merge`.

It would also be nice if a missing value was represented as a Conflict, so that the type could be

```
/// Either a legacy tree in a `TreeValue::Tree`
/// or a conflict containing legacy trees.
LegacyTree(&'a TreeValue),
Merge(Conflict<Option<TreeValue>>)
```